docs/alerts.md

```diff
@@ -0,0 +1,183 @@
+
+Each of these levels will be represented by an integer to make it easier to compare the levels.
+There will also be an additional level to represent that a check was not able to determine the status or the status was not yet known: `unknown`.
+
+They will be mapped as the following:
+
+|level|value|
+|-----|-----|
+|ok|0|
+|info|1|
+|warn|2|
+|crit|3|
+|unknown|-1|
+
+### Check
+
+The `check` function will take an input and evaluate each row to determine the level.
+The function will have a predicate for each specified level and will output a level for the highest matching check.
+One output table will be produced for each input table.
+The output tables will have the same schema, but with one additional column that is `_level:int`.
+
+`check` has the following properties:
+
+| Name | Type                | Description
+| ---- | ----                | -----------
+| crit | (r: record) -> bool | The function that determines if at the `crit` level. If unset, it is ignored.
+| warn | (r: record) -> bool | The function that determines if at the `warn` level. If unset, it is ignored.
+| info | (r: record) -> bool | The function that determines if at the `info` level. If unset, it is ignored.
+| ok   | (r: record) -> bool | The function that determines if at the `ok` level. If unset, this will always return true.
+
+If multiple of the above are true, then the highest level wins.
+If none of the above are true, then `-1 (unknown)` is the result.
+The methods above will short-circuit so there is no expectation that all of them will be invoked for every row.
+The functions should not have side effects.
+
+Example 1:
+
+```
+from(bucket: "telegraf")
+    |> range(start: -5m)
+    |> filter(fn: (r) => r._measurement == "cpu" and r._field == "usage_system")
+    |> check(
+    	crit: (r) => r._value > 90,
+    	warn: (r) => r._value > 80,
+    )
+```
+
+|Input:|_time|_value|Output:|_time|_value|_level|
+|------|-----|------|-------|-----|------|------|
+| |0|45| |0|45|ok|
+| |10|55| |10|55|ok|
+| |20|95| |20|95|crit|
+| |30|92| |30|92|crit|
+| |40|89| |40|89|warn|
+| |50|91| |50|91|crit|
+| |60|87| |60|87|warn|
+| |70|81| |70|81|warn|
+| |80|50| |80|50|ok|
+| |90|60| |90|60|ok|
+| |100|30| |100|30|ok|
+| |110|25| |110|25|ok|
+| |120|10| |120|10|ok|
+
+Example 2:
+
+```
+from(bucket: "telegraf")
+    |> range(start: -5m)
+    |> filter(fn: (r) => r._measurement == "cpu" and r._field == "usage_system")
+    |> check(
+    	crit: (r) => r._value > 90,
+    	warn: (r) => r._value > 80,
+    	// Do not reset to the ok level until the system usage drops enough.
+    	ok:   (r) => r._value <= 20,
+    )
+```
+
+|Input:|_time|_value|Output:|_time|_value|_level|
+|------|-----|------|-------|-----|------|------|
+| |0|45| |0|45|ok|
+| |10|55| |10|55|ok|
+| |20|95| |20|95|crit|
+| |30|92| |30|92|crit|
+| |40|89| |40|89|warn|
+| |50|91| |50|91|crit|
+| |60|87| |60|87|warn|
+| |70|81| |70|81|warn|
+| |80|50| |80|50|unknown|
```

Comment:
don't you think it would be better to keep the last valid state instead of `unknown`?

Reply:
There's a few reasons why not:

1. We may not have access to that. This doesn't assume that it knows the value of the last row so it is stateless and relies on the handler to determine what it should do with the input instead.
2. It would be a lie to have it give the last valid state. It's not in the warning or critical state anymore, so why should we report that it is?

Reply:
I agree with you, only wanted to see the cons :)